### PR TITLE
bazel: generate a single coverage binary for all C++ tests.

### PR DESCRIPTION
**Pull request by htuch**
This patch has a number of changes to how the build system works and how tests with non-trivial
environment setups get their files:

* Eliminate setup_cmds and envoy_cc_test_with_json. These setup scripts are now run where necessary
  inside the test fixture or equivalent functionality is now in TestEnvironment.

* Provide a single test binary target for all C++ (but not shell) tests in
  //test/coverage:coverage_tests. This is needed for reasonable performance coverage under Bazel,
  see https://github.com/bazelbuild/bazel/issues/1118 for why. Unfortunately, we need to generate
  the BUILD for this manually, I have a script in test/coverage/gen_build.sh to get this to happen.

* Avoid false dependency on //test/integration:integration_lib by test/main.cc. This came up while
  working on the coverage performance. Small tests, e.g. hex_test, were dragging in the integration
  test library for no reason, just because of some logging setup in main.cc.

* Set alwayslink=1 on all C++ libraries. While working on the above, strange link dependency orders
  were causing a problem. These mostly pertained to the places I had broken circular link
  dependencies, e.g. //source/common/upstream:upstream_{lib,includes}. There's cleaner ways to fix
  this than alwayslink=1, but I think that we probably do want to link all object files that the Envoy
  binary depends on and alwayslink=1 will prevent surprises for non-Bazel experts (e.g. when adding a
  new filter library).

### CODE REVIEW

File: test/coverage/BUILD [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# This file is generated by test/coverage/gen_build.sh. DO NOT EDIT. Rerun
```
Comment (mattklein123) [inline test/coverage/BUILD]: This feels extremely fragile and nearly guaranteed to fall out of date. Is there any way that test libraries can somehow register themselves to be part of an optional monolithic test process. Furthermore, is there any reason for this even to be coverage specific? I.e., can we just have an option to build a monolithic test binary regardless of whether we are doing ASAN, coverage, etc.? This seems like potentially the way to go in CI where as I've mentioned elsewhere, the extra links don't seem to provide much value?

Comment (htuch) [inline test/coverage/BUILD]: Bazel supports tagging, we use it in this patch to opt tests into the monolithic build via gen_build.sh and its bazel query. Unfortunately, dependencies can't make use of tagging information when we assemble a binary, see https://github.com/bazelbuild/bazel/issues/2362.

Here's what I propose to avoid the fragility:

1. We don't check in test/coverage/BUILD at all.
2. We have the `tests/run_envoy_coverage.sh` script invoke `test/coverage/gen_build.sh` prior to the build. The resulting build file is ephemeral and we don't have stale BUILD files in the repository.

Does that sound acceptable?

For ASAN and other testing, I think we should try and avoid a monolithic build. The reason for this is that when we switch to the next generation CI, there will be the ability to treat individual tests targets independently as first-class resources, with the ability to dive into their history and see them on a test grid. If you prefer we put them in a single binary for Travis CI performance until we go there, please let me know and I can make this non-coverage specific.

Comment (mattklein123) [inline test/coverage/BUILD]: No that's fine. This plan sounds good (generate BUILD file during coverage run). +1

File: test/BUILD [hunk 2]
```diff
@@ -1,6 +1,6 @@
 package(default_visibility = ["//visibility:public"])
 
-load("//bazel:envoy_build_system.bzl", "envoy_cc_test_library")
+load("//bazel:envoy_build_system.bzl", "envoy_cc_test", "envoy_cc_test_library")
```
Comment (mattklein123) [inline test/BUILD]: Is this change needed if there are no other changes in this file? (Mostly trying to understand bazel).

File: test/coverage/gen_build.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+BUILD_PATH="$(dirname "$0")"/BUILD
+
+TARGETS=$("${BAZEL_BIN}" query "attr('tags', 'coverage_test_lib', //test/...)")
+(
+  cat << EOF
+# This file is generated by test/coverage/gen_build.sh automatically prior to
+# coverage runs. It is under .gitignore. DO NOT EDIT, DO NOT CHECK IN.
+package(default_visibility = ["//visibility:public"])
+
+load("//bazel:envoy_build_system.bzl", "envoy_cc_test", "envoy_cc_test_library")
```
Comment (mattklein123) [inline test/coverage/gen_build.sh]: same here, do we need "envoy_cc_test_library"? It's not used.

File: test/integration/integration.cc [hunk 4]
```diff
@@ -212,8 +211,8 @@ void IntegrationTcpClient::ConnectionCallbacks::onEvent(uint32_t events) {
 
 BaseIntegrationTest::BaseIntegrationTest()
     : api_(new Api::Impl(std::chrono::milliseconds(10000))),
-      dispatcher_(api_->allocateDispatcher()) {
-
+      dispatcher_(api_->allocateDispatcher()),
+      default_log_level_(TestEnvironment::getOptions().logLevel()) {
```
Comment (mattklein123) [inline test/integration/integration.cc]: I'm a little confused why this whole getOptions() dance is needed (referring to all of the code in this PR that sets up the static). I spend a few minutes looking, but can you maybe just summarize?

File: test/BUILD (hunk 2 above)
Comment (htuch) [inline test/BUILD]: No, that was just something from an earlier experiment.

File: test/coverage/gen_build.sh (hunk 3 above)
Comment (htuch) [inline test/coverage/gen_build.sh]: Copy+paste snafu.

File: test/integration/integration.cc (hunk 4 above)
Comment (htuch) [inline test/integration/integration.cc]: I didn't want `//test:main` to take a dependency on `//test/integration:integration_lib`, so rather than initialize the library in `main.cc`, we just stash away the args for later when used by integration tests. This then avoids building `integration_lib` (and consequently, most of Envoy), when running tiny tests like `//test/common/common:hex_test`.

Comment (mattklein123) [inline test/integration/integration.cc]: Got it, makes sense. Thanks.
